### ISSUE TITLE
Strange scope issue when use lock (cause deadlock)

### ISSUE THREAD

**Issue by TheWaWaR** [rust-lang/rust#60107]
### The code:
``` rust
use std::sync::Mutex;

#[derive(Clone, Debug)]
struct A {}

fn main() {
    let n = Mutex::new(Some(A {}));
    
    if let Some(a) = { let a = n.lock().unwrap(); a.clone() } {
        n.lock();
    }
    
    println!("This is ok");
    
    if let Some(a) = { let a = n.lock(); a.unwrap().clone() } {
        n.lock();
    }

    println!("Never reached !!!!");
}
```

### Output:
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.56s
     Running `target/debug/playground`
/root/entrypoint.sh: line 8:     7 Killed                  timeout --signal=KILL ${timeout} "$@"
Standard Output
This is ok
```

**Comment by jonas-schievink** [comment]
Pretty sure this is expected due to the way scopes in match statements work - the matched expression is dropped only after the match completes

**Comment by pnkfelix** [comment]
@jonas-schievink well, what you said, plus the detail about how the temporaries in tail expressions of a block are given the dynamic extent of the block's parent, right?

as demonstrated here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a08de70a5bf898ae5daa2a5a2f538955

cc #46413
cc #37612

**Comment by jonas-schievink** [comment]
Yes, indeed. Looks like https://github.com/rust-lang/rust/issues/37612#issuecomment-304626464 explains this quite well, and that issue is basically the same as this (minus the `if let` -> `match` desugaring), so I'll close this as a duplicate of https://github.com/rust-lang/rust/issues/37612.
